### PR TITLE
TRUNK-4639 ""equals(Object obj)" should be overridden along with the "co...

### PR DESCRIPTION
**Pull request by tomaszmueller**
...mpareTo(T obj)" method"

### CODE REVIEW

**Comment by tomaszmueller** [comment]
@rkorytkowski https://issues.openmrs.org/browse/TRUNK-4639

**Comment by tomaszmueller** [comment]
@rkorytkowski changed accordingly to comment in https://issues.openmrs.org/browse/TRUNK-4639

File: api/src/main/java/org/openmrs/PersonName.java [hunk 1]
```diff
@@ -530,42 +531,12 @@ public String toString() {
 	 * @should return negative if other familynamePrefix is greater
 	 * @should return negative if other familyNameSuffix is greater
 	 * @should return negative if other dateCreated is greater
+	 * @Depracated since 1.12. Use DefaultCompartor instead.
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/PersonName.java]: misspelled DefaultComparator :)

File: web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java [hunk 2]
```diff
@@ -387,6 +387,42 @@ public int compareTo(FieldHolder other) {
...
+					if (other.getGroupMemberConcepts() != null) {
+						return false;
+					}
+				} else if (!groupMemberConcepts.equals(other.getGroupMemberConcepts())) {
+					return false;
+				}
+			} // end else
+			
+			return true;
+		} //
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java]: Whenever you implement equals you must implement hashCode. Also read JavaDoc for compareTo. If equals returns true then compareTo must return 0. If equals returns false then compareTo must return != 0. You need to correct equals to behave accordingly. In other words it's enough to implement equals as return compareTo() == 0.

File: api/src/main/java/org/openmrs/PersonName.java (hunk 1 above)
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/PersonName.java]: Could we please add in addition to @Deprecated the following:
Note: this comparator imposes orderings that are inconsistent with equals.

Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/PersonName.java]: Please do the same for all other compareTo methods.

**Comment by tomaszmueller** [comment]
@rkorytkowski done

File: web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java [hunk 3]
```diff
@@ -387,6 +387,25 @@ public int compareTo(FieldHolder other) {
...
+		public boolean equals(FieldHolder other) {
+			return compareTo(other) == 0;
+		} //
+		
+		/**
+		 * Returns a hash code based on the formField fields used in comparing.
+		 * @see java.lang.Object#hashCode()
+		 */
+		public int hashCode() {
+			return formField.getPageNumber() * formField.getFieldNumber()
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java]: The rule is that hashCodes must be the same if equals returns true. It's not the case here if pageNumber is equal and some other field is different. The easiest fix here is to include just the first field that is checked in compareTo i.e. pageNumber. Please use HashCodeBuilder to generate the value.

**Comment by tomaszmueller** [comment]
Fixed hashCode

File: web/pom.xml [hunk 4]
```diff
@@ -136,6 +136,11 @@
 		 <artifactId>mysql-connector-mxj</artifactId>
 		 <scope>provided</scope>
 	  </dependency>
+       <dependency>
```
Comment (rkorytkowski) [inline web/pom.xml]: Please remove this dependency.

File: web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java [hunk 5]
```diff
@@ -387,6 +388,23 @@ public int compareTo(FieldHolder other) {
...
+		public boolean equals(FieldHolder other) {
+			return compareTo(other) == 0;
+		} //
+		
+		/**
+		 * Returns a hash code based on the formField.getPageNumber field used in comparing.
+		 * @see java.lang.Object#hashCode()
+		 */
+		public int hashCode() {
+			return HashCodeBuilder.reflectionHashCode(formField.getPageNumber());
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/controller/encounter/EncounterDisplayController.java]: This should call .append(formField.getPageNumber()).build();

**Comment by dkayiwa** [comment]
@tomaszmueller can you squash these commits into one?

**Comment by tomaszmueller** [comment]
@dkayiwa  squashed

**Comment by rkorytkowski** [comment]
Travis reports a failing build. The following test fails: saveShortPatient_shouldAddANewPersonAttributeWithANonEmptyValue(org.openmrs.web.controller.patient.ShortPatientFormControllerTest)
@tomaszmueller, does it pass for you locally?

**Comment by tomaszmueller** [comment]
@rkorytkowski It didn't pass
I fixed it and updated the pull request

File: web/pom.xml [hunk 6]
```diff
@@ -131,11 +131,6 @@
 		 <groupId>org.apache.velocity</groupId>
 		 <artifactId>velocity-tools</artifactId>
 	  </dependency>
-	  <dependency>
```
Comment (rkorytkowski) [inline web/pom.xml]: Why did you remove this?

Comment (tomaszmueller) [inline web/pom.xml]: @rkorytkowski Didn't you asked to remove this dependency in one of your comments on outdated diff 8 days ago? Or did I mistunderstood which dependency should I remove?

Comment (rkorytkowski) [inline web/pom.xml]: It was some other dependency that you included by mistake.

Comment (tomaszmueller) [inline web/pom.xml]: @rkorytkowski added the dependency again

**Comment by rkorytkowski** [comment]
Well done!
